src/access/AccessRestrictionProvider.h

```diff
@@ -0,0 +1,283 @@
+     */
+    struct Restriction
+    {
+        Type restrictionType;
+        Optional<uint32_t> id;
+    };
+
+    /**
+     * Defines a single entry in the access restriction list, which contains a list of restrictions
+     * for a cluster on an endpoint.
+     */
+    struct Entry
+    {
+        FabricIndex fabricIndex;
+        EndpointId endpointNumber;
+        ClusterId clusterId;
+        std::vector<Restriction> restrictions;
+    };
+
+    /**
+     * Defines the interface for a checker for access restriction exceptions.
+     */
+    class AccessRestrictionExceptionChecker
+    {
+        public:
+            virtual ~AccessRestrictionExceptionChecker() = default;
+
+            /**
+             * Check if any restrictions are allowed to be applied to the given request.
+             *
+             * @retval true if restrictions may NOT be applied
+             */
+            virtual bool AreRestrictionsDisallowed(const SubjectDescriptor & subjectDescriptor,
+                                                   const RequestPath & requestPath) = 0;
+    };
+
+    /**
+     * Define a standard implementation of the AccessRestrictionExceptionChecker interface
+     * which is the default implementation used by AccessResrictionProvider.
+     */
+    class StandardAccessRestrictionExceptionChecker : public AccessRestrictionExceptionChecker
+    {
+        public:
+            StandardAccessRestrictionExceptionChecker() = default;
+            ~StandardAccessRestrictionExceptionChecker() = default;
+
+            /**
+             * Check if any restrictions are allowed to be applied to the given request.
+             *
+             * @retval true if restrictions may NOT be applied
+             */
+            bool AreRestrictionsDisallowed(const SubjectDescriptor & subjectDescriptor,
+                                           const RequestPath & requestPath) override;
+    };
+
+    /**
+     * Used to notify of changes in the access restriction list and active reviews.
+     */
+    class Listener
+    {
+    public:
+        virtual ~Listener() = default;
+
+        /**
+         * Notifies of a change in the commissioning access restriction list.
+         */
+        virtual void CommissioningRestrictionListChanged() = 0;
+
+        /**
+         * Notifies of a change in the access restriction list.
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the list has changed.
+         */
+        virtual void RestrictionListChanged(FabricIndex fabricIndex) = 0;
+
+        /**
+         * Notifies of an update to an active review with instructions and an optional redirect URL.
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the entry has changed.
+         * @param [in] token        The token of the review being updated (obtained from ReviewFabricRestrictionsResponse)
+         * @param [in] instruction  The instructions to be displayed to the user.
```

Comment:
This is optional so should maybe be allowed null as well.

Reply:
done (now Optional Span)